torch/lib/THD/base/data_channels/DataChannelGloo.hpp

```diff
@@ -17,17 +23,190 @@ namespace std {
+      hash<thd::rank_type>()(std::get<5>(k))
+    );
   }
 };
 
 } // namespace std
 
 namespace thd {
 
+struct GlooCache;
+
+struct DataChannelGloo : DataChannel {
+  using store_type = ::gloo::rendezvous::Store;
+
+
+  struct RequestGloo : DataChannel::Request {
+    RequestGloo(QueueWorker::Request&& request);
+    virtual ~RequestGloo();
+
+    virtual bool isCompleted() override;
+    virtual void wait() override;
+
+  private:
+    QueueWorker::Request _request;
+  };
+
+  DataChannelGloo();
+  DataChannelGloo(int timeout);
+  virtual ~DataChannelGloo();
+
+  bool init() override;
+
+  rank_type getRank() override;
+  rank_type getNumProcesses() override;
+
+  void allGather(std::vector<thpp::Tensor*>& output, thpp::Tensor& input,
+                 THDGroup group_id = THDGroupWORLD) override;
+  void gather(std::vector<thpp::Tensor*>& output, thpp::Tensor& input,
+              rank_type dst_rank, THDGroup group_id = THDGroupWORLD) override;
+  void scatter(std::vector<thpp::Tensor*>& input, thpp::Tensor& output,
+               rank_type src_rank, THDGroup group_id = THDGroupWORLD) override;
+  void allReduce(thpp::Tensor& data, THDReduceOp operation,
+                 THDGroup group_id = THDGroupWORLD) override;
+  void reduce(thpp::Tensor& data, THDReduceOp operation, rank_type dst_rank,
+              THDGroup group_id = THDGroupWORLD) override;
+  void broadcast(thpp::Tensor& data, rank_type src_id,
+                 THDGroup group_id = THDGroupWORLD) override;
+  void send(const Scalar& data, rank_type dst_id) override;
+  void send(thpp::Tensor& data, rank_type dst_id) override;
+  void receive(Scalar& data, rank_type src_id) override;
+  void receive(thpp::Tensor& data) override;
+  void receive(thpp::Tensor& data, rank_type src_id) override;
+  RequestGloo* isend(thpp::Tensor& data, rank_type dst_rank) override;
+  RequestGloo* ireceive(thpp::Tensor& data, rank_type src_rank) override;
+
+  void barrier(THDGroup group_id = THDGroupWORLD) override;
+
+  THDGroup newGroup(const std::vector<rank_type>& ranks) override;
+
+private:
+  template<typename T>
+  void allGatherT(std::vector<thpp::Tensor*>& output,
+                  thpp::Tensor& input, THDGroup group_id);
+
+  template<typename T>
+  void allReduceT(thpp::Tensor& data, THDReduceOp operation,
+                  THDGroup group_id = THDGroupWORLD);
+
+  template<typename T>
+  void broadcastT(thpp::Tensor& data, rank_type src_rank,
+                  THDGroup group_id = THDGroupWORLD);
+
+  rank_type _rank; // Current process' rank
+  std::string _addr;
+  port_type _port;
+  rank_type _num_processes; // Number of processes in network
+  std::shared_ptr<store_type> _store;
+  std::shared_ptr<::gloo::transport::Device> _device;
+  std::unordered_map<THDGroup, DataChannel::Group> _groups;
+  
+  std::unique_ptr<GlooCache> _cache;
+
+  // Workers
+  QueueWorker _send_worker, _receive_worker;
+};
+
+
+template<DataOperation D, typename T>
```

Comment:
Eh I know it's a bit late, but can you please change `DataOperation` to a different name (e.g. `CollectiveType` - this could even be defined somewhere in `base`. It seems as if it could be a useful enum helper.